mlflow/projects/__init__.py

```diff
@@ -93,41 +85,26 @@ def _resolve_experiment_id(experiment_name=None, experiment_id=None):
 
 
 def _run(uri, experiment_id, entry_point="main", version=None, parameters=None,
-         backend=None, backend_config=None, use_conda=True,
+         backend_name=None, backend_config=None, use_conda=True,
          storage_dir=None, synchronous=True, run_id=None):
     """
     Helper that delegates to the project-running method corresponding to the passed-in backend.
     Returns a ``SubmittedRun`` corresponding to the project run.
     """
-
-    parameters = parameters or {}
-    work_dir = _fetch_project(uri=uri, force_tempdir=False, version=version)
-    project = _project_spec.load_project(work_dir)
-    _validate_execution_environment(project, backend)
+    project, work_dir = fetch_and_validate_project(uri, version, entry_point, parameters)
+    _validate_execution_environment(project, backend_name)
     project.get_entry_point(entry_point)._validate_parameters(parameters)
-    if run_id:
-        active_run = tracking.MlflowClient().get_run(run_id)
-    else:
-        active_run = _create_run(uri, experiment_id, work_dir, entry_point)
-
-    # Consolidate parameters for logging.
-    # `storage_dir` is `None` since we want to log actual path not downloaded local path
-    entry_point_obj = project.get_entry_point(entry_point)
-    final_params, extra_params = entry_point_obj.compute_parameters(parameters, storage_dir=None)
-    for key, value in (list(final_params.items()) + list(extra_params.items())):
-        tracking.MlflowClient().log_param(active_run.info.run_id, key, value)
-
-    repo_url = _get_git_repo_url(work_dir)
-    if repo_url is not None:
-        for tag in [MLFLOW_GIT_REPO_URL, LEGACY_MLFLOW_GIT_REPO_URL]:
-            tracking.MlflowClient().set_tag(active_run.info.run_id, tag, repo_url)
-
-    # Add branch name tag if a branch is specified through -version
-    if _is_valid_branch_name(work_dir, version):
-        for tag in [MLFLOW_GIT_BRANCH, LEGACY_MLFLOW_GIT_BRANCH_NAME]:
-            tracking.MlflowClient().set_tag(active_run.info.run_id, tag, version)
-
-    if backend == "databricks":
+    active_run = get_or_create_run(run_id, uri, experiment_id, work_dir, entry_point)
+    log_project_params_and_tags(active_run, project, work_dir, entry_point, parameters, version)
```

Comment:
In a similar vein to my question further down about whether the MLflow client or the project backend should be responsible for MLflow run creation. I'm also wondering whether the project backend or the client should be responsible for logging project parameters, the working directory, etc. 

For example, this approach seems to require the client to always fetch the project in order to determine the working directory name. It seems feasible for backends to resolve projects directly in the future, independently of the client (e.g., this would probably be the case if we support the execution of Databricks notebooks as MLflow projects). @smurching do you have thoughts here?

Reply:
Given the discussion in https://github.com/mlflow/mlflow/pull/2566#discussion_r394840637, I think it's fair to delegate these responsibilities to the MLflow client, as is done here.

Reply:
As an outcome of discussion here: https://github.com/mlflow/mlflow/pull/2566/files#r394638164, I think we agree that client run creation makes sense. However, we also agree that project backends should manage the resolution of projects; this presents a bit of a wrinkle with respect to the tag populated by `work_dir`, which isn't determined until the project is fetched. Perhaps backends should set this tag instead, leaving everything else to the client?

Reply:
This method `log_project_params_and_tags` is supposed to be called by project backends. It is stilled called in `_run` method to keep the previous behavior for databricks, local and k8s backends.
It will be removed as soon they backends will implement the new interfaces (in a next PR)

FYI I will implement also @smurching's suggestion to remove this method and to give access to the methods that create runs, fetch the project & validate the config, etc.